File: solutions/problems/linkedLists/mergeKSortedLists.ts
/**
 * Approach: a simple binary min-heap (array + sift, keyed by node value)
 * holding each list's current node. Repeatedly pop the smallest, append
 * it to the result, and push its successor.
 * Time: O(N log k), N = total nodes, k = number of lists. Space: O(k).
 */

import { ListNode } from "../../common/listNode";

export function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
  const heap: ListNode[] = [];

  const siftUp = (i: number) => {
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      if (heap[i].val < heap[parent].val) {
        [heap[i], heap[parent]] = [heap[parent], heap[i]];
        i = parent;
      } else break;
    }
  };
  const siftDown = (i: number) => {
    const n = heap.length;
    while (true) {
      const l = 2 * i + 1;
      const r = 2 * i + 2;
      let smallest = i;
      if (l < n && heap[l].val < heap[smallest].val) smallest = l;
      if (r < n && heap[r].val < heap[smallest].val) smallest = r;
      if (smallest === i) break;
      [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
      i = smallest;
    }
  };
  const push = (node: ListNode) => {
    heap.push(node);
    siftUp(heap.length - 1);
  };
  const pop = (): ListNode => {
    const top = heap[0];
    const last = heap.pop()!;
    if (heap.length > 0) {
      heap[0] = last;
      siftDown(0);
    }
    return top;
  };

  for (const node of lists) if (node) push(node);

  const dummy = new ListNode();
  let tail = dummy;
  while (heap.length > 0) {
    const node = pop();
    tail.next = node;
    tail = tail.next;
    if (node.next) push(node.next);
  }
  return dummy.next;
}

File: problems/linked_lists/merge_k_sorted_lists.py
"""
LeetCode 23. Merge k Sorted Lists

Merge k sorted linked lists into one sorted list.

Example:
    [[1,4,5],[1,3,4],[2,6]] -> [1,1,2,3,4,4,5,6]

Approach: push each list's current node into a min-heap keyed by value
(with a tie-breaking counter, since ListNode isn't orderable). Repeatedly
pop the smallest, append it to the result, and push its successor.

Time:  O(N log k), N = total nodes, k = number of lists
Space: O(k) for the heap
"""

import heapq
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))
from common.list_node import ListNode, build_linked_list, linked_list_to_list


def merge_k_lists(lists):
    heap = []
    counter = 0  # tie-breaker so heapq never compares ListNode objects
    for node in lists:
        if node:
            heapq.heappush(heap, (node.val, counter, node))
            counter += 1

    dummy = ListNode()
    tail = dummy
    while heap:
        _, _, node = heapq.heappop(heap)
        tail.next = node
        tail = tail.next
        if node.next:
            heapq.heappush(heap, (node.next.val, counter, node.next))
            counter += 1

    return dummy.next


if __name__ == "__main__":
    lists = [build_linked_list(l) for l in ([1, 4, 5], [1, 3, 4], [2, 6])]
    merged = merge_k_lists(lists)
    assert linked_list_to_list(merged) == [1, 1, 2, 3, 4, 4, 5, 6]

    assert merge_k_lists([]) is None
    assert merge_k_lists([None]) is None
    print("All tests passed.")
